api/ombpdf/semdb.py

```diff
@@ -1,12 +1,153 @@
+        footnote_citation.footnote_node = footnote_citation.footnote_node
+        footnote_citation.save()
+
+    return writer.cursor
+
+
+class DatabaseWriter(semtree.Writer):
+    def __init__(self, policy):
+        self.cursor_stack = [PDFAwareCursor.new_tree('policy', policy=policy)]
+        self.footnote_citations = {}
+
+    @property
+    def cursor(self):
+        return self.cursor_stack[-1]
+
+    @property
+    def sec_level(self):
+        return self.cursor.identifier.split('_').count('sec')
+
+    def end_element(self, el):
+        self.cursor.model.text = self.cursor.text.rstrip()
+        super().end_element(el)
+
+    def begin_document(self, doc):
+        pass
+
+    def end_document(self, doc):
+        self.cursor_stack = self.cursor_stack[:1]
+
+    def begin_heading(self, heading):
+        while heading.level <= self.sec_level:
+            self.cursor_stack.pop()
+        while heading.level > self.sec_level:
+            self.cursor_stack.append(self.cursor.add_child('sec'))
+        self.cursor_stack.append(self.cursor.add_child('heading'))
+        self.cursor.pdf_node = heading
+
+    def end_heading(self, heading):
+        text = self.cursor.text.strip()
+        self.cursor_stack.pop()
+        max_length = DocNode._meta.get_field('title').max_length
+        if len(text) > max_length:
+            logger.warning('Heading too long. Misparse? %s', text)
+        self.cursor.model.title = text[:max_length]
+
+    def begin_paragraph(self, p):
+        self.cursor_stack.append(self.cursor.add_child('para'))
+        self.cursor.pdf_node = p
+
+    def end_paragraph(self, p):
+        self.cursor_stack.pop()
+
+    def begin_list(self, li):
+        self.cursor_stack.append(self.cursor.add_child('list'))
+        self.cursor.pdf_node = li
+
+    def end_list(self, li):
+        self.cursor_stack.pop()
+
+    def begin_list_item(self, p):
+        li = self.cursor.add_child('listitem')
+        self.cursor_stack.append(li.add_child('para'))
+        self.cursor.pdf_node = p
+
+        parent_list = next(
+            self.cursor.ancestors(lambda n: n.node_type == 'list'),
+            None,
+        )
+        if not parent_list:
+            logger.warning('Adding list item, but no list parent')
+            li.model.marker = '?'
+        elif parent_list.pdf_node.is_ordered:
+            li.model.marker = f'{li.model.type_emblem}.'
+        else:
+            li.model.marker = '\u2022'
+
+    def end_list_item(self, p):
+        self.cursor_stack.pop()
+
+    def begin_footnote_list(self, fl):
+        self.cursor_stack = self.cursor_stack[:1]
+
+    def end_footnote_list(self, fl):
+        no_nodes = [key for key, value in self.footnote_citations.items()
+                    if not value.footnote_node]
+        if no_nodes:
+            logger.warning('Unresolved footnote citations exist: %s', no_nodes)
+            for no_node_cite in no_nodes:
+                del self.footnote_citations[no_node_cite]
```

Comment:
Hmm, do you think we should modify our DB schema to support unresolved footnote citations if the policy is still in some kind of draft/unpublished state?  my concern is that this removal of information could result in user confusion--e.g., a document with an unresolved footnote citation for "4" will likely be easier to fix than a document that just skips from footnote 3 to 5 without any mention of a "4" at all.

Anyways, I guess this question might be outside the scope of this PR, since leaving it unaddressed doesn't actually lose us any functionality (I assume the current XML intermediary solution has the same problem).

Reply:
I think I agree with you in general, that adding more meta data around parse errors would be useful. We can probably grep around for warning log messages and turn them into some sort of annotation. I also agree that that's a bit outside of scope for this PR.

In this particular case, I want to emphasize we're dropping the citation annotation rather than the footnote itself. The text for the citation will still exist, and if its corresponding footnote (perhaps misparsed) is somewhere in the document, it'd still be placed at the bottom of the document. I *think* that's low risk at this point; the information we're losing is either a false positive or it'll be relatively easy to reconstruct in the editor.